subprojects/base-services/src/main/java/org/gradle/internal/work/DefaultWorkerLeaseService.java

```diff
@@ -0,0 +1,327 @@
+    private final ProjectLockRegistry projectLockRegistry;
+    private final WorkerLeaseLockRegistry workerLeaseLockRegistry;
+
+    public DefaultWorkerLeaseService(ResourceLockCoordinationService coordinationService, boolean parallelEnabled, int maxWorkerCount) {
+        this.maxWorkerCount = maxWorkerCount;
+        this.coordinationService = coordinationService;
+        this.projectLockRegistry = new ProjectLockRegistry(coordinationService, parallelEnabled);
+        this.workerLeaseLockRegistry = new WorkerLeaseLockRegistry(coordinationService);
+        LOGGER.info("Using {} worker leases.", maxWorkerCount);
+    }
+
+    @Override
+    public WorkerLease getCurrentWorkerLease() {
+        Collection<? extends ResourceLock> operations = workerLeaseLockRegistry.getResourceLocksByCurrentThread();
+        if (operations.isEmpty()) {
+            throw new IllegalStateException("No worker lease associated with the current thread");
+        }
+        return (DefaultWorkerLease) operations.toArray()[operations.size() - 1];
+    }
+
+    private synchronized DefaultWorkerLease getWorkerLease(LeaseHolder parent) {
+        int workerId = counter++;
+        Thread ownerThread = Thread.currentThread();
+        return workerLeaseLockRegistry.getResourceLock(parent, workerId, ownerThread);
+    }
+
+    @Override
+    public DefaultWorkerLease getWorkerLease() {
+        Collection<? extends ResourceLock> operations = workerLeaseLockRegistry.getResourceLocksByCurrentThread();
+        LeaseHolder parent = operations.isEmpty() ? root : (DefaultWorkerLease) operations.toArray()[operations.size() - 1];
+        return getWorkerLease(parent);
+    }
+
+    @Override
+    public void stop() {
+        coordinationService.withStateLock(new Transformer<ResourceLockState.Disposition, ResourceLockState>() {
+            @Override
+            public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
+                if (workerLeaseLockRegistry.hasOpenLocks()) {
+                    throw new IllegalStateException("Some worker leases have not been marked as completed.");
+                }
+                if (projectLockRegistry.hasOpenLocks()) {
+                    throw new IllegalStateException("Some project locks have not been unlocked.");
+                }
+                return FINISHED;
+            }
+        });
+    }
+
+    @Override
+    public ResourceLock getProjectLock(String gradlePath, String projectPath) {
+        return projectLockRegistry.getResourceLock(gradlePath, projectPath);
+    }
+
+    @Override
+    public void withoutProjectLock(Runnable runnable) {
+        final Collection<? extends ResourceLock> projectLocks = projectLockRegistry.getResourceLocksByCurrentThread();
+        withoutLocks(projectLocks.toArray(new ResourceLock[projectLocks.size()])).execute(runnable);
+    }
+
+    @Override
+    public Executor withLocks(final ResourceLock... locks) {
```

Comment:
Yeah, I wasn't super comfortable with this.  The _only_ reason I did it this way is that `withLocks(Runnable, ResourceLock...)` seemed even more awkward and executor pretty much matched the interface I was looking for.  Good point on the returning a value, though. 
 I'll change it to something like `<T> T withLocks(Iterable<ResourceLock>, Callable<T>)`.

I've captured this in a follow up story.